build: Qt 5.15.x

fanquake opened the pull request:
Opening this for visibility. Apart from it probably being time to use a newer version of Qt for releases, an update is required so that we are using a version which has a C++17 mode.

There's a lot of commits here, because I've made some changes (like rebasing our patches), that aren't strictly necessary. There are also some changes, like bumping the minimum required macOS version, that will likely happen separate, as part of the C++17 migration. So this change will end up being a bit smaller.

The biggest change here is that [Qt no longer has internal xcb libs](https://codereview.qt.nokia.com/c/qt/qtbase/+/253905), so we need to build the requirements ourselves. This also means some additional build-time tools, such as `yacc` (from `bison`) are also required; to build xkbcommon.

I've tested this on Ubuntu; although it'd be great if more people could test with their particular brand of Linux desktop.

<img width="568" alt="Linux Qt" src="https://user-images.githubusercontent.com/863730/90225425-ee469680-de43-11ea-8c12-849421c4497f.png">

Cross-compiled Windows and macOS builds compile and run, but the GUI currently looks a bit broken on both. i.e macOS seems to look a bit like an older version of Windows:

<img width="1040" alt="Screen Shot 2020-08-11 at 8 47 00 pm" src="https://user-images.githubusercontent.com/863730/90225973-d91e3780-de44-11ea-9bc8-5895cb0287e4.png">

I haven't had enough time to fully investigate (also why there's some WIP commits), so if anyone wants to dig in and see what Qt has changed since 5.9.x feel free.

fanquake committed:
depends: remove fix_configure_mac.patch from qt

Fixed upstream in Qt 5.11.0RC1

https://bugreports.qt.io/browse/QTBUG-67286

fanquake committed:
depends: remove fix_riscv64_arch.patch from qt

Was fixed upstream in 6a39e49a6cdeb28a04a3657bb6a22f848d5dfa9d

fanquake committed:
depends: remove fix_rcc_determinism.patch from qt

Fixed upstream in https://bugreports.qt.io/browse/QTBUG-62511

fanquake committed:
depends: remove xkb-default.patch from qt

This was removed upstream in d5abf545971da717014d316127045fc19edbcd65

fanquake committed:
build: qt 5.15.2

fanquake committed:
build: update fix_pkg_config.patch for qt 5.15

fanquake committed:
build: update fix_no_printer.patch for qt 5.15

fanquake committed:
build: remove fix_android_qmake_conf.patch from qt build

This may need to be reinstated in some capacity, but I haven't looked
through the changes here, and have no ability to test them.

fanquake committed:
build: update no-xlib.patch for qt 5.15

fanquake committed:
build: update fix_android_jni_static.patch for qt 5.15

fanquake committed:
build: xproto 7.0.31

fanquake committed:
build: libXau 1.0.9

fanquake committed:
build: xcb_proto 1.14

fanquake committed:
build: libxcb 1.14

Minimum required libxcb to build qt 5.15.x is 1.11.

https://codereview.qt.nokia.com/c/qt/qtbase/+/253905

Some plugins have been re-enabled as they are required by Qt.

fanquake committed:
build: freetype 2.10.4

Co-authored-by: mammix2 <mammix2@hotmail.com>

fanquake committed:
build: add xkbcommon 0.8.4

Currently 0.8.4 because at 0.9.0 they switch from autotools to meson..

This also requires yacc (bison) to be available at build time.

fanquake committed:
build: add libxcb_util 0.4.0

fanquake committed:
build: add libxcb_util_render 0.3.9

fanquake committed:
build: add libxcb_util_keysyms 0.4.0

fanquake committed:
build: add libxcb_util_image 0.4.0

fanquake committed:
build: patch around broken pkg-config in libxcb_util_image

fanquake committed:
build: add libxcb_util_wm 0.4.1

Required for xcb_icccm.

fanquake committed:
build: no-longer disable dark mode on macOS

We should no-longer have to disable darkmode, as Qt should
properly support it, and we are using a new enough SDK.

fanquake committed:
build: expat 2.2.10

MarcoFalke commented:
Added 0.22 milestone based on the assumption that none of this is required to build with a qt5.15 system package? (Only depends changes here?)

fanquake commented:
> the assumption that none of this is required to build with a qt5.15 system package?

As far as I'm aware people have already been building with Qt 5.15, and that has been working after #19097.

jonasschnelli commented:
Thanks!
It's currently failing on gitian: https://bitcoin.jonasschnelli.ch/gitian/builds/231/build_osx.log

hebasto commented:
@fanquake 
> Cross-compiled Windows and macOS builds compile and run, but the GUI currently looks a bit broken on both. i.e macOS seems to look a bit like an older version of Windows:

I found this https://github.com/hebasto/bitcoin/commit/528360ade9bb3f4902b632f6f755cf63b3397412 useful for diagnostics such issues. Feel free to integrate :)

promag commented:
Nice work! I already used Qt 5.15 and I don't think it looked like that. Will check this soon™

vasild commented:
There are a bunch of compilation warnings on `master` and on this PR with Qt 5.15.0 like:

```
src/qt/transactionview.cpp:587:13: warning: 'QDateTime' is deprecated: Use QDate::startOfDay() [-Wdeprecated-declarations]
            QDateTime(dateTo->date()).addDays(1));
            ^
```

Here is a fix:

<details>
<summary>Fix -Wdeprecated-declarations warnings with Qt 5.15</summary>

```diff
diff --git i/src/qt/sendcoinsdialog.cpp w/src/qt/sendcoinsdialog.cpp
index 97fb88d71..786af1a73 100644
--- i/src/qt/sendcoinsdialog.cpp
+++ w/src/qt/sendcoinsdialog.cpp
@@ -168,14 +168,14 @@ void SendCoinsDialog::setModel(WalletModel *_model)
         // fee section
         for (const int n : confTargets) {
             ui->confTargetSelector->addItem(tr("%1 (%2 blocks)").arg(GUIUtil::formatNiceTimeOffset(n*Params().GetConsensus().nPowTargetSpacing)).arg(n));
         }
         connect(ui->confTargetSelector, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &SendCoinsDialog::updateSmartFeeLabel);
         connect(ui->confTargetSelector, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &SendCoinsDialog::coinControlUpdateLabels);
-        connect(ui->groupFee, static_cast<void (QButtonGroup::*)(int)>(&QButtonGroup::buttonClicked), this, &SendCoinsDialog::updateFeeSectionControls);
-        connect(ui->groupFee, static_cast<void (QButtonGroup::*)(int)>(&QButtonGroup::buttonClicked), this, &SendCoinsDialog::coinControlUpdateLabels);
+        connect(ui->groupFee, &QButtonGroup::idClicked, this, &SendCoinsDialog::updateFeeSectionControls);
+        connect(ui->groupFee, &QButtonGroup::idClicked, this, &SendCoinsDialog::coinControlUpdateLabels);
         connect(ui->customFee, &BitcoinAmountField::valueChanged, this, &SendCoinsDialog::coinControlUpdateLabels);
         connect(ui->optInRBF, &QCheckBox::stateChanged, this, &SendCoinsDialog::updateSmartFeeLabel);
         connect(ui->optInRBF, &QCheckBox::stateChanged, this, &SendCoinsDialog::coinControlUpdateLabels);
         CAmount requiredFee = model->wallet().getRequiredFee(1000);
         ui->customFee->SetMinValue(requiredFee);
         if (ui->customFee->value() < requiredFee) {
diff --git i/src/qt/transactionview.cpp w/src/qt/transactionview.cpp
index 54ecfc38e..5f638f559 100644
--- i/src/qt/transactionview.cpp
+++ w/src/qt/transactionview.cpp
@@ -234,13 +234,13 @@ void TransactionView::setModel(WalletModel *_model)
 
         columnResizingFixer = new GUIUtil::TableViewLastColumnResizingFixer(transactionView, AMOUNT_MINIMUM_COLUMN_WIDTH, MINIMUM_COLUMN_WIDTH, this);
 
         if (_model->getOptionsModel())
         {
             // Add third party transaction URLs to context menu
-            QStringList listUrls = _model->getOptionsModel()->getThirdPartyTxUrls().split("|", QString::SkipEmptyParts);
+            QStringList listUrls = _model->getOptionsModel()->getThirdPartyTxUrls().split("|", Qt::SkipEmptyParts);
             for (int i = 0; i < listUrls.size(); ++i)
             {
                 QString url = listUrls[i].trimmed();
                 QString host = QUrl(url, QUrl::StrictMode).host();
                 if (!host.isEmpty())
                 {
@@ -272,36 +272,36 @@ void TransactionView::chooseDate(int idx)
         transactionProxyModel->setDateRange(
                 TransactionFilterProxy::MIN_DATE,
                 TransactionFilterProxy::MAX_DATE);
         break;
     case Today:
         transactionProxyModel->setDateRange(
-                QDateTime(current),
+                current.startOfDay(),
                 TransactionFilterProxy::MAX_DATE);
         break;
     case ThisWeek: {
         // Find last Monday
         QDate startOfWeek = current.addDays(-(current.dayOfWeek()-1));
         transactionProxyModel->setDateRange(
-                QDateTime(startOfWeek),
+                startOfWeek.startOfDay(),
                 TransactionFilterProxy::MAX_DATE);
 
         } break;
     case ThisMonth:
         transactionProxyModel->setDateRange(
-                QDateTime(QDate(current.year(), current.month(), 1)),
+                QDate(current.year(), current.month(), 1).startOfDay(),
                 TransactionFilterProxy::MAX_DATE);
         break;
     case LastMonth:
         transactionProxyModel->setDateRange(
-                QDateTime(QDate(current.year(), current.month(), 1).addMonths(-1)),
-                QDateTime(QDate(current.year(), current.month(), 1)));
+                QDate(current.year(), current.month(), 1).addMonths(-1).startOfDay(),
+                QDate(current.year(), current.month(), 1).startOfDay());
         break;
     case ThisYear:
         transactionProxyModel->setDateRange(
-                QDateTime(QDate(current.year(), 1, 1)),
+                QDate(current.year(), 1, 1).startOfDay(),
                 TransactionFilterProxy::MAX_DATE);
         break;
     case Range:
         dateRangeWidget->setVisible(true);
         dateRangeChanged();
         break;
@@ -580,14 +580,14 @@ QWidget *TransactionView::createDateRangeWidget()
 
 void TransactionView::dateRangeChanged()
 {
     if(!transactionProxyModel)
         return;
     transactionProxyModel->setDateRange(
-            QDateTime(dateFrom->date()),
-            QDateTime(dateTo->date()).addDays(1));
+            dateFrom->date().startOfDay(),
+            dateTo->date().startOfDay().addDays(1));
 }
 
 void TransactionView::focusTransaction(const QModelIndex &idx)
 {
     if(!transactionProxyModel)
         return;
```
</details>

[qt_5.15_fix_deprecation_warnings.diff.txt](https://github.com/bitcoin/bitcoin/files/5096419/qt_5.15_fix_deprecation_warnings.diff.txt)

fanquake commented:
@vasild Thanks. See https://github.com/bitcoin-core/gui/pull/46 which should take care of all of them.

fanquake commented:
I've fixed up a bunch of things here.
Rebased for #19689.
Dropped the [WIP] qt commits and merged the relevant changes in to the qt 5.15.x commit.
Modified the libxcb_util_image fix to be a patch file rather than using sed.
Added a commit that should be enough to get the CI running.
Fixed the docs for the dependency changes.

Note that Qt 5.15.1 has been tagged (no downloads yet), so once that's up I'll swap over.

Also dumped some output using @hebasto's patch. Turns out the macOS UI looks like Windows because the Windows style is being used.. Unclear why this is happening. Will investigate. Debug output:
```bash
2020-08-24T11:22:47.398810Z [main] Bitcoin Core version v0.20.99.0-1971aead8-dirty (release build)
2020-08-24T11:22:47.398842Z [main] Qt 5.15.0 (static), plugin=cocoa (static)
2020-08-24T11:22:47.398936Z [main] Static Plugin: {"IID":"org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3","MetaData":{"Keys":["cocoa"]},"archreq":0,"className":"QCocoaIntegrationPlugin","debug":false,"version":331520}
2020-08-24T11:22:47.398959Z [main] Style=windows / QWindowsStyle
2020-08-24T11:22:47.398973Z [main] System: macOS 10.15, x86_64-little_endian-lp64
2020-08-24T11:22:47.398987Z [main] Screen: Color LCD 1680x1050, pixel ratio=2.0
2020-08-24T11:22:47.398997Z [main] Screen: LG HDR 4K 2560x1440, pixel ratio=2.0
2020-08-24T11:22:47.958407Z [main] GUI: Populating font family aliases took 512 ms. Replace uses of missing font family ".AppleSystemUIFont,13,-1,5,50,0,0,0,0,0" with one that exists to avoid this cost. 
2020-08-24T11:22:48.816413Z [main] GUI: initializeResult : Initialization result:  true
2020-08-24T11:22:48.821713Z [main] GUI: Platform customization: "macosx"
```

hebasto commented on ci/test/00_setup_env.sh:
```diff
@@ -65,7 +65,7 @@ export BASE_OUTDIR=${BASE_OUTDIR:-$BASE_SCRATCH_DIR/out/$HOST}
 export PREVIOUS_RELEASES_DIR=${PREVIOUS_RELEASES_DIR:-$BASE_ROOT_DIR/releases/$HOST}
 export SDK_URL=${SDK_URL:-https://bitcoincore.org/depends-sources/sdks}
-export DOCKER_PACKAGES=${DOCKER_PACKAGES:-build-essential libtool autotools-dev automake pkg-config bsdmainutils curl ca-certificates ccache python3 rsync git procps}
+export DOCKER_PACKAGES=${DOCKER_PACKAGES:-build-essential bison libtool autotools-dev automake pkg-config bsdmainutils curl ca-certificates ccache python3 rsync git procps}
```
1e17e1038cfcc8b0769917a8d9107eac27889601

To fix this error in CentOS build
```
Building libxkbcommon...
make[1]: Entering directory `/home/travis/build/hebasto/bitcoin/depends/work/build/i686-pc-linux-gnu/libxkbcommon/0.8.4-b76cb6131e8'
  YACC     src/xkbcomp/parser.c
./build-aux/ylwrap: line 176: yacc: command not found
```

this change is required
```diff
--- a/ci/test/00_setup_env_i686_centos.sh
+++ b/ci/test/00_setup_env_i686_centos.sh
@@ -9,7 +9,7 @@ export LC_ALL=C.UTF-8
 export HOST=i686-pc-linux-gnu
 export CONTAINER_NAME=ci_i686_centos_7
 export DOCKER_NAME_TAG=centos:7
-export DOCKER_PACKAGES="gcc-c++ glibc-devel.x86_64 libstdc++-devel.x86_64 glibc-devel.i686 libstdc++-devel.i686 ccache libtool make git python3 python36-zmq which patch lbzip2 dash"
+export DOCKER_PACKAGES="gcc-c++ glibc-devel.x86_64 libstdc++-devel.x86_64 glibc-devel.i686 libstdc++-devel.i686 ccache libtool make git python3 python36-zmq which patch lbzip2 dash bison"
 export GOAL="install"
 export BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-reduce-exports --with-boost-process"
 export CONFIG_SHELL="/bin/dash"
```

hebasto commented:
It seems the CentOS build is a subject of https://bugreports.qt.io/browse/QTBUG-85669

hebasto commented on depends/packages/qt.mk:
```diff
@@ -24,13 +24,14 @@ define $(package)_set_vars
 $(package)_config_opts_debug = -debug
 $(package)_config_opts += -bindir $(build_prefix)/bin
-$(package)_config_opts += -c++std c++11
+$(package)_config_opts += -c++std c++17
```
Qt configure reports an error on Xenial build:
```
ERROR: Feature 'c++1z' was enabled, but the pre-condition 'features.c++17' failed.
```

fanquake replied:
Ubuntu 16.04 is not listed as [supported by Qt 5.15.x](https://doc.qt.io/qt-5/supported-platforms.html). Regardless, if Qts build system cant detect that your compiler supports C++17 it's not going to build in C++17 mode. Looking at the GCC on Xenial, it's [5.3.1](https://packages.ubuntu.com/xenial/gcc), [which only has support for some C++17 features](https://gcc.gnu.org/projects/cxx-status.html#cxx17).

If you post the actual precondition that failed (it'll be earlier in the log) we could take a look at it.

hebasto replied:
> If you post the actual precondition that failed (it'll be earlier in the log) we could take a look at it.

https://travis-ci.org/github/hebasto/bitcoin/jobs/720916965

hebasto replied:
```
Checking for C++17 support... 
+ cd /home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/config.tests/cxx17 && /home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/bin/qmake "CONFIG -= qt debug_and_release app_bundle lib_bundle" "CONFIG += static warn_off console single_arch" -early "CONFIG += cross_compile" /home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/config.tests/cxx17
+ cd /home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/config.tests/cxx17 && MAKEFLAGS= /usr/bin/make
> make[1]: Entering directory '/home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/config.tests/cxx17'
> g++ -c -m64 -pipe -pipe -O2 -I/home/travis/build/hebasto/bitcoin/depends/x86_64-pc-linux-gnu/include -O2 -std=gnu++1z -w -fPIC  -I. -I/home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/mkspecs/linux-g++-64 -o main.o main.cpp
> main.cpp:8:19: fatal error: variant: No such file or directory
> compilation terminated.
> Makefile:171: recipe for target 'main.o' failed
> make[1]: *** [main.o] Error 1
> make[1]: Leaving directory '/home/travis/build/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/qt/5.15.0-e4e26f06c7b/qtbase/config.tests/cxx17'
test config.qtbase.tests.cxx17 FAILE
```

fanquake replied:
> fatal error: variant: No such file or directory

The problem is that the compiler doesn't support C++17. GCC didn't introduce support for `std::variant` (in libstdc++) until the [7.1 release](https://gcc.gnu.org/gcc-7/changes.html). Xenial ships with 5.3.1. So when Qts build system tests for C++17 support by checking it can compile something using `std::variant`, it's failing as expected.

MarcoFalke commented:
This is currently assigned to the 0.22 milestone, by which time both centos:7 and xenial are no longer supported. So no need to test on them right now.

hebasto commented:
> This is currently assigned to the 0.22 milestone, by which time both centos:7 and xenial are no longer supported. So no need to test on them right now.

So let's remove them form CI jobs in this PR.

hebasto commented:
macOS 10.12 build error (on Travis):
```
Project ERROR: Cannot run target compiler 'x86_64-apple-darwin16-clang++'. Output:
===================
===================
Maybe you forgot to setup the environment?
```

hebasto commented:
Windows build on Travis is a subject of https://forum.qt.io/topic/113250/building-qt-fails-with-missing-dxgil_3-h

hebasto commented:
It would be nice to add a doc commit with the minimum MinGW version requirement:
- https://doc.qt.io/qt-5/windows-requirements.html#gcc-mingw-builds

hebasto commented:
> Windows build on Travis is a subject of https://forum.qt.io/topic/113250/building-qt-fails-with-missing-dxgil_3-h

Could be fixed by bumping Ubuntu version up to 20.04:
```diff
--- a/ci/test/00_setup_env_win64.sh
+++ b/ci/test/00_setup_env_win64.sh
@@ -7,7 +7,7 @@
 export LC_ALL=C.UTF-8
 
 export CONTAINER_NAME=ci_win64
-export DOCKER_NAME_TAG=ubuntu:18.04  # Check that bionic can cross-compile to win64 (bionic is used in the gitian build as well)
+export DOCKER_NAME_TAG=ubuntu:20.04
 export HOST=x86_64-w64-mingw32
 export PACKAGES="python3 nsis g++-mingw-w64-x86-64 wine-binfmt wine64"
 export RUN_FUNCTIONAL_TESTS=false
```

MarcoFalke commented:
> So let's remove them from CI jobs in this PR.

They won't get removed. #19504 is the first pull that goes into 0.22 and that simply bumps them: https://github.com/bitcoin/bitcoin/pull/19504/files#r455507731

hebasto commented:
From the Qt [source](https://github.com/qt/qtbase/blob/ba3b53cb501a77144aa6259e48a8e0edc3d1481d/src/platformsupport/platformsupport.pro#L19-L20) it follows that this change is required:
```diff
--- a/build-aux/m4/bitcoin_qt.m4
+++ b/build-aux/m4/bitcoin_qt.m4
@@ -309,8 +309,8 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
       PKG_CHECK_MODULES([QTFB], [Qt5FbSupport], [QT_LIBS="-lQt5FbSupport $QT_LIBS"])
       PKG_CHECK_MODULES([QTFONTDATABASE], [Qt5FontDatabaseSupport], [QT_LIBS="-lQt5FontDatabaseSupport $QT_LIBS"])
       PKG_CHECK_MODULES([QTTHEME], [Qt5ThemeSupport], [QT_LIBS="-lQt5ThemeSupport $QT_LIBS"])
-      PKG_CHECK_MODULES([QTSERVICE], [Qt5ServiceSupport], [QT_LIBS="-lQt5ServiceSupport $QT_LIBS"])
       if test "x$TARGET_OS" = xlinux; then
+        PKG_CHECK_MODULES([QTSERVICE], [Qt5ServiceSupport], [QT_LIBS="-lQt5ServiceSupport $QT_LIBS"])
         PKG_CHECK_MODULES([QTXCBQPA], [Qt5XcbQpa], [QT_LIBS="$QTXCBQPA_LIBS $QT_LIBS"])
       elif test "x$TARGET_OS" = xdarwin; then
         PKG_CHECK_MODULES([QTCLIPBOARD], [Qt5ClipboardSupport], [QT_LIBS="-lQt5ClipboardSupport $QT_LIBS"])
```

At least, it fixes the Windows build on Travis :)

fanquake committed:
build: remove freetype_back_compat.patch for qt 5.15

By the time we ship a release with Qt 5.15, we'll certainly no-longer be
supporting Ubuntu 14.04 and Ubuntu 16.04 ships with FreeType 2.6.1,
which is new enough that using the symbol is no-longer an issue.

The renaming of FT_Get_X11_Font_Format() happened in FreeType 2.6

fanquake commented:
Rebased after #19761.
Fixed-up `QTSERVICE`, thanks @hebasto.
Dropped the freetype back-compat patch. See commit message for reasoning.
Turned off some additional Qt features, like the markdown reader and SQL item models. We also don't use GSS but `no-gss` doesn't seem to work.

Some more TODO:
Figure out how much more of our patching / hacking we can drop.
I'm also planning on splitting anything C++17-bump related into a separate PR, so it's clear what will be going in before this PR is merged, and what isn't required testing wise. i.e building this assumes the use of a C++17 capable compiler.

hebasto commented:
Tested 3f4fe21d3fa04cc8f284dccae03a5e4a27dba9e4, still having:
- on [Windows build](https://travis-ci.org/github/hebasto/bitcoin/jobs/721663770#L4475):
```
In file included from rhi/qrhi.cpp:49:0:
rhi/qrhid3d11_p_p.h:57:10: fatal error: dxgi1_3.h: No such file or directory
 #include <dxgi1_3.h>
          ^~~~~~~~~~~
compilation terminated.
```

- on [macOS build](https://travis-ci.org/github/hebasto/bitcoin/jobs/721663771#L6340):
```
Project ERROR: Cannot run target compiler 'x86_64-apple-darwin16-clang++'. Output:
===================
===================
Maybe you forgot to setup the environment?
```

hebasto commented:
I've managed to fix Windows build with the following patch:
```diff
--- a/qtbase/src/gui/gui.pro
+++ b/qtbase/src/gui/gui.pro
@@ -49,7 +49,6 @@
 include(itemmodels/itemmodels.pri)
 include(vulkan/vulkan.pri)
 include(platform/platform.pri)
-include(rhi/rhi.pri)
 
 QMAKE_LIBS += $$QMAKE_LIBS_GUI
 
```

It opts out a new [RHI functionality](https://www.qt.io/blog/qt-roadmap-for-2020), and, as a side effect, reduces compile time.

Branch: https://github.com/hebasto/bitcoin/commits/pr19716-0827-win
Travis: https://travis-ci.org/github/hebasto/bitcoin/builds/721800296

fanquake commented:
> I've managed to fix Windows build

Is this fixing a problem other than `dxgi1_3.h: No such file or directory` mentioned above? In that case the problem, and the solution is pretty clear. The CI is using a version of the mingw headers that are too old to build Qt. This will be fixed when we use a newer version of Ubuntu for builds.

I'm also not sure that disabling RHI is the right approach, as it seems that's what Qt is going to be using for all it's graphics and rendering from Qt 6 on-wards. Also, the blog post you linked says " For Qt 5.15 LTS RHI is an opt-in functionality", so is a patch actually required to disable it? Shouldn't there be (a combination of) configure flags that will opt you in/out of the functionality?

hebasto commented:
> Is this fixing a problem other than `dxgi1_3.h: No such file or directory` mentioned above?

Yes, it is.

> In that case the problem, and the solution is pretty clear. The CI is using a version of the mingw headers that are too old to build Qt. This will be fixed when we use a newer version of Ubuntu for builds.

Sure, https://github.com/bitcoin/bitcoin/pull/19716#issuecomment-679964477

> I'm also not sure that disabling RHI is the right approach, as it seems that's what Qt is going to be using for all it's graphics and rendering from Qt 6 on-wards. Also, the blog post you linked says " For Qt 5.15 LTS RHI is an opt-in functionality", so is a patch actually required to disable it? Shouldn't there be (a combination of) configure flags that will opt you in/out of the functionality?

I have not found any `configure` options or features to disable RHI. By the time we switch to Qt 6 we will use a newer version of Ubuntu for builds for sure :)

hebasto commented:
IIUC, RHI is used for rendering in Qt Quick framework, not in Qt Widgets.

MarcoFalke commented:
There will probably be fewer warning by bumping the qt version, so you might be able to remove the `Temporary workaround for https://github.com/bitcoin/bitcoin/issues/16368`

fanquake commented:
Rebased, and updated the win64 CI now that #18921 has been merged.

hebasto commented:
>  - on [macOS build](https://travis-ci.org/github/hebasto/bitcoin/jobs/721663771#L6340):
> 
> ```
> Project ERROR: Cannot run target compiler 'x86_64-apple-darwin16-clang++'. Output:
> ===================
> ===================
> Maybe you forgot to setup the environment?
> ```

@fanquake 
This problem could be solved with https://github.com/hebasto/bitcoin/commit/cc5752f6788ee8408f98ce19c91a7fe0771d5d20

But another problem with SDK arises:
```
> Unknown error resolving current platform SDK version.
```

fanquake committed:
build: remove fix_mingw_cross_compile.patch for qt 5.15

fanquake commented:
Rebased on top of #19867. This now also builds Qt 5.15.1.

> @fanquake
This problem could be solved with hebasto@cc5752f

@hebasto I think the correct solution here is actually to update our [sed commands that perform a similar replacement](https://github.com/bitcoin/bitcoin/blob/a33651866c863328e8aca870465cc1c99a4240d6/depends/packages/qt.mk#L224).

fanquake committed:
build: remove qt libpng powerpc patch

fanquake commented:
Rebased on master. Updated some dependencies. Updated to Qt 5.15.2.  I'll continue splitting a few more changes off here.

jonasschnelli commented:
I think this might be helpful for the macOS11 visual glitches https://github.com/bitcoin-core/gui/issues/136.
But maybe its still too early,.. see https://bugreports.qt.io/browse/QTBUG-85546.

hebasto reviewed (commented):
3975ec0a5f5a298b93e7e39397b44602105ad4b7

On CentOS 8:
```
$ make -C depends qt
...
Building libxkbcommon...
make[1]: Entering directory '/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9'
  YACC     src/xkbcomp/parser.c
yacc: e - line 219 of "/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9/src/xkbcomp/parser.y", syntax error
%destructor { FreeStmt((ParseCommon *) $$); }
^
make[1]: *** [Makefile:1705: src/xkbcomp/parser.c] Error 1
make[1]: Leaving directory '/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9'
make: *** [funcs.mk:267: /home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9/./.stamp_built] Error 2
make: Leaving directory '/home/hebasto/bitcoin/depends'
```

BlockMechanic commented:
> [3975ec0](https://github.com/bitcoin/bitcoin/commit/3975ec0a5f5a298b93e7e39397b44602105ad4b7)
> 
> On CentOS 8:
> 
> ```
> $ make -C depends qt
> ...
> Building libxkbcommon...
> make[1]: Entering directory '/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9'
>   YACC     src/xkbcomp/parser.c
> yacc: e - line 219 of "/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9/src/xkbcomp/parser.y", syntax error
> %destructor { FreeStmt((ParseCommon *) $$); }
> ^
> make[1]: *** [Makefile:1705: src/xkbcomp/parser.c] Error 1
> make[1]: Leaving directory '/home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9'
> make: *** [funcs.mk:267: /home/hebasto/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libxkbcommon/0.8.4-fd57d1b36f9/./.stamp_built] Error 2
> make: Leaving directory '/home/hebasto/bitcoin/depends'
> ```

 Same issue on ubuntu 18.04

BlockMechanic commented:
There is a problem with libxkbcommon ,  some reading :- https://stackoverflow.com/questions/50794051/ubuntu-byacc-btyacc-syntax-error. I ended up just installing bison

fanquake committed:
build: only pass -optimized-tools to qt in debug mode

Qt's configure tells us that "-optimized-tools is not useful in -release
mode.", so don't use it there.

fanquake commented:
> The issue with with libxkbcommon is fixed by installing bison rather than byacc

The CI changes to install bison have now been merged into the "qt 5.15.2" commit, rather than being in a separate WIP commit. Also added bb38e8a30de2278b4bd9dae6f80cc097892dd5c8 to stop using `-optimized-tools` in release mode. We can split off more commits if need be.

While this still has at least one outstanding issue, which is fixing the macOS cross-compile, I'm going to take it out of draft. If anyone wants to help debug why the macOS theming (when built using locally using depends) isn't correct, that'd be handy. It also needs more testing across OSs, particularly Linux.

fanquake committed:
build: disable qt SDK version checking

This tries to invoke xcrun, which is not available when cross-compiling.
Given we are in control of the SDK versions being used, removing this
check has minimal-no effect.

fanquake commented:
I've now fixed the macOS cross-compilation issue and added a commit that disables some SDK version checking, as that would try and invoke `xcrun`, which isn't available when cross-compiling. I've also added a commit (262e02995912a9caa4026105d9ce6237965e8600) which might help debug the theme issues.

jarolrod commented:
Any consensus surrounding Qt dropping LTS for `Qt 5.15`? 
See: https://www.theregister.com/2021/01/05/qt_lts_goes_commercial_only/

laanwj commented:
> Any consensus surrounding Qt dropping LTS for Qt 5.15?

I don't think that's too relevant here. It means that upstream won't do any updates on top of 5.15. But we generally don't track minor versions of Qt so for us it makes no difference. The next Qt bump after this is likely to be 6.x.

laanwj commented:
I'm confused by this Win64 CI error:
```
make[4]: Leaving directory '/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-w64-mingw32/src'
Running tests: allocator_tests from test/allocator_tests.cpp
wine: fixed a temporary directory error, please try running the same command again
make[3]: *** [Makefile:16848: test/addrman_tests.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
wine: fixed a temporary directory error, please try running the same command again
make[3]: *** [Makefile:16848: test/allocator_tests.cpp.test] Error 1
wine: created the configuration directory '/root/.wine'
wine: chdir to /tmp/wine-4SkHOY/server-32-27828a : No such file or directory
wine: chdir to /tmp/wine-2CwwWn/server-32-27828a : No such file or directory
make[3]: *** [Makefile:16848: test/amount_tests.cpp.test] Error 1
make[3]: *** [Makefile:16848: test/arith_uint256_tests.cpp.test] Error 1
make[3]: Leaving directory '/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-w64-mingw32/src'
make[2]: *** [Makefile:15367: check-am] Error 2
make[2]: Leaving directory '/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-w64-mingw32/src'
make[1]: *** [Makefile:15052: check-recursive] Error 1
make[1]: Leaving directory '/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-w64-mingw32/src'
make: *** [Makefile:813: check-recursive] Error 1
```

MarcoFalke commented:
Looks like an upstream bug. Locally I could fix it by running the command twice, IIRC

MarcoFalke commented on ci/test/00_setup_env_win64.sh:
```diff
@@ -7,7 +7,7 @@
 
 export CONTAINER_NAME=ci_win64
-export DOCKER_NAME_TAG=ubuntu:18.04  # Check that bionic can cross-compile to win64 (bionic is used in the gitian build as well)
+export DOCKER_NAME_TAG=ubuntu:20.04
```
I don't like this change, but without it the ci fails: https://cirrus-ci.com/task/6121023374360576?command=ci#L5504

```
compiling .rcc/release/qrc_qmake_webgradients.cpp
In file included from rhi/qrhi.cpp:49:0:
rhi/qrhid3d11_p_p.h:57:10: fatal error: dxgi1_3.h: No such file or directory
 #include <dxgi1_3.h>
          ^~~~~~~~~~~
compilation terminated.
Makefile.Release:71663: recipe for target '.obj/release/qrhi.o' failed
make[3]: *** [.obj/release/qrhi.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from rhi/qrhid3d11.cpp:37:0:
rhi/qrhid3d11_p_p.h:57:10: fatal error: dxgi1_3.h: No such file or directory
 #include <dxgi1_3.h>
          ^~~~~~~~~~~
compilation terminated.
Makefile.Release:72684: recipe for target '.obj/release/qrhid3d11.o' failed
make[3]: *** [.obj/release/qrhid3d11.o] Error 1
make[3]: Leaving directory '/tmp/cirrus-ci-build/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-102170435c3/qtbase/src/gui'
Makefile:45: recipe for target 'release' failed
make[2]: *** [release] Error 2
make[2]: Leaving directory '/tmp/cirrus-ci-build/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-102170435c3/qtbase/src/gui'
Makefile:423: recipe for target 'sub-gui' failed
make[1]: *** [sub-gui] Error 2
make[1]: Leaving directory '/tmp/cirrus-ci-build/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-102170435c3/qtbase/src'
funcs.mk:273: recipe for target '/tmp/cirrus-ci-build/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-102170435c3/./.stamp_built' failed
make: *** [/tmp/cirrus-ci-build/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-102170435c3/./.stamp_built] Error 2
make: Leaving directory '/tmp/cirrus-ci-build/depends'
Exit status: 2

MarcoFalke replied:
Do the gitian builds pass? If yes, why do they and the ci doesn't?

fanquake replied:
> I don't like this change, but without it the ci fails:

A newer version of mingw-w64 & headers is required to build this version of Qt, so we (essentially) have to use a newer version of Ubuntu in the CI & gitian (I'll add those changes, but will PR them separate in advance). This was also discussed above.

MarcoFalke replied:
Note to myself: https://github.com/bitcoin/bitcoin/pull/19716#issuecomment-767347016

MarcoFalke commented:
```
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-apple-darwin18'
+ make -j1 -C src check-security
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-apple-darwin18/src'
Checking binary security...
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-apple-darwin18/src'
+ make -j1 -C src check-symbols
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-apple-darwin18/src'
Checking macOS dynamic libraries...
Security is not in ALLOWED_LIBRARIES!
Metal is not in ALLOWED_LIBRARIES!
QuartzCore is not in ALLOWED_LIBRARIES!
IOSurface is not in ALLOWED_LIBRARIES!
ColorSync is not in ALLOWED_LIBRARIES!
CoreVideo is not in ALLOWED_LIBRARIES!
qt/bitcoin-qt: failed DYNAMIC_LIBRARIES
Makefile:14994: recipe for target 'check-symbols' failed
make: *** [check-symbols] Error 1
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-apple-darwin18/src'
```

```
compiling rhi/qrhi.cpp
In file included from rhi/qrhi.cpp:49:0:
rhi/qrhid3d11_p_p.h:57:10: fatal error: dxgi1_3.h: No such file or directory
 #include <dxgi1_3.h>
          ^~~~~~~~~~~
compilation terminated.
Makefile.Release:71663: recipe for target '.obj/release/qrhi.o' failed
make[3]: *** [.obj/release/qrhi.o] Error 1
make[3]: Leaving directory '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-b4b0c80531f/qtbase/src/gui'
Makefile:45: recipe for target 'release' failed
make[2]: *** [release] Error 2
make[2]: Leaving directory '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-b4b0c80531f/qtbase/src/gui'
Makefile:423: recipe for target 'sub-gui' failed
make[1]: *** [sub-gui] Error 2
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-b4b0c80531f/qtbase/src'
funcs.mk:273: recipe for target '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-b4b0c80531f/./.stamp_built' failed
make: *** [/home/ubuntu/build/bitcoin/depends/work/build/x86_64-w64-mingw32/qt/5.15.2-b4b0c80531f/./.stamp_built] Error 2
make: Leaving directory '/home/ubuntu/build/bitcoin/depends'
```

```
Building libxkbcommon...
make[1]: Entering directory '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-linux-gnu/libxkbcommon/0.8.4-68d09a7117d'
  YACC     src/xkbcomp/parser.c
./build-aux/ylwrap: line 176: yacc: command not found
Makefile:1705: recipe for target 'src/xkbcomp/parser.c' failed
make[1]: *** [src/xkbcomp/parser.c] Error 127
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-linux-gnu/libxkbcommon/0.8.4-68d09a7117d'
funcs.mk:273: recipe for target '/home/ubuntu/build/bitcoin/depends/work/build/x86_64-linux-gnu/libxkbcommon/0.8.4-68d09a7117d/./.stamp_built' failed
make: *** [/home/ubuntu/build/bitcoin/depends/work/build/x86_64-linux-gnu/libxkbcommon/0.8.4-68d09a7117d/./.stamp_built] Error 2
make: Leaving directory '/home/ubuntu/build/bitcoin/depends'

MarcoFalke commented:
```
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32'
+ make -j1 -C src check-security
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/src'
Checking binary security...
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/src'
+ make -j1 -C src check-symbols
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/src'
Checking Windows dynamic libraries...
NETAPI32.dll is not in ALLOWED_LIBRARIES!
ole32.dll is not in ALLOWED_LIBRARIES!
qt/bitcoin-qt.exe: failed DYNAMIC_LIBRARIES
make: *** [Makefile:16867: check-symbols] Error 1
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/src'

fanquake commented:
Going to propose some smaller, simpler changes. New PR as they are substantially different from what is here.
